### PR TITLE
pkg/pdutil(ticdc): improve pd time accuracy

### PR DESCRIPTION
**Pull request by overvenus**
### What problem does this PR solve?

Issue Number: ref #4757

### What is changed and how it works?

Add duration to ts event time.

#### Tests 

 - Unit test

##### Will it cause performance regression or break compatibility?

No.

##### Do you need to update user documentation, design documentation or monitoring documentation?

No.

### Release note 

```release-note
None
```

### CODE REVIEW

File: pkg/pdutil/clock.go [hunk 1]
```diff
@@ -41,32 +41,38 @@ type clock struct {
...
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	ret.mu.timeCache = oracle.GetTimeFromTS(oracle.ComposeTS(physical, 0))
+	ret.mu.tsEventTime = oracle.GetTimeFromTS(oracle.ComposeTS(physical, 0))
+	ret.mu.tsProcessingTime = time.Now()
 	return ret, nil
 }
 
 // Run will get time from pd periodically to cache in timeCache
```
Comment (asddongmen) [inline pkg/pdutil/clock.go:71]: `timeCache` in the comment should be remove.

File: pkg/pdutil/clock.go [hunk 2]
```diff
@@ -100,10 +109,10 @@ func (c *clock) Run(ctx context.Context) {
 // CurrentTime returns current time from timeCache
```
Comment (asddongmen) [inline pkg/pdutil/clock.go:109]: ditto

File: pkg/pdutil/clock.go [hunk 3]
```diff
@@ -100,10 +109,10 @@ func (c *clock) Run(ctx context.Context) {
...
 func (c *clock) CurrentTime() (time.Time, error) {
 	c.mu.RLock()
-	err := c.mu.err
-	cacheTime := c.mu.timeCache
-	c.mu.RUnlock()
-	return cacheTime, errors.Trace(err)
+	defer c.mu.RUnlock()
+	tsEventTime := c.mu.tsEventTime
+	current := tsEventTime.Add(time.Since(c.mu.tsProcessingTime))
```
Comment (asddongmen) [inline pkg/pdutil/clock.go:114]: why can we just return `tsEventTime.Add(time.Since(c.mu.tsProcessingTime))`?

Comment (overvenus) [reply pkg/pdutil/clock.go:114]: To make code easier to understand

**Review by 3AceShowHand — APPROVED** [review]

**Review by asddongmen — APPROVED** [review]

File: pkg/pdutil/clock.go [hunk 4]
```diff
@@ -41,32 +41,38 @@ type clock struct {
...
+	ret.mu.tsEventTime = oracle.GetTimeFromTS(oracle.ComposeTS(physical, 0))
+	ret.mu.tsProcessingTime = time.Now()
 	return ret, nil
 }
 
-// Run will get time from pd periodically to cache in timeCache
+// Run gets time from pd periodically.
 func (c *clock) Run(ctx context.Context) {
 	ctx, cancel := context.WithCancel(ctx)
 	c.cancel = cancel
```
Comment (3AceShowHand) [inline pkg/pdutil/clock.go:74]: `cancel` is called by `stop`, cause race detected in the test.

File: pkg/pdutil/clock.go [hunk 5]
```diff
@@ -81,33 +89,38 @@ func (c *clock) Run(ctx context.Context) {
...
+	tsEventTime := c.mu.tsEventTime
+	current := tsEventTime.Add(time.Since(c.mu.tsProcessingTime))
+	return current, errors.Trace(c.mu.err)
 }
 
 // Stop clock.
 func (c *clock) Stop() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
 	c.cancel()
```
Comment (3AceShowHand) [inline pkg/pdutil/clock.go:124]: it looks that the `cancel` maybe nil?

**Pull request merged**